src/main/java/com/iota/iri/network/neighbor/impl/NeighborImpl.java

```diff
@@ -0,0 +1,274 @@
+                    metrics.incrInvalidProtocolMessageLengthCount();
+
+                    // abort connection if we are handshaking
+                    if (state == NeighborState.HANDSHAKING) {
+                        return -1;
+                    }
+                    return bytesRead;
+                }
+
+                // if we are handshaking, then we must have a handshaking packet as the initial packet
+                if (state == NeighborState.HANDSHAKING
+                        && protocolHeader.getMessageType() != ProtocolMessage.HANDSHAKE) {
+                    log.error("neighbor {}'s initial packet is not a handshaking packet, closing connection",
+                            getHostAddressAndPort());
+                    return -1;
+                }
+
+                // we got the header, now we want to read the message
+                readState = ReadState.READ_MESSAGE;
+                msgReader = MessageReaderFactory.create(protocolHeader.getMessageType(),
+                        protocolHeader.getMessageLength());
+                // execute another read as we likely already have the message in the network buffer
+                return read();
+
+            case READ_MESSAGE:
+                switch (msgReader.getMessageType()) {
+                    case HANDSHAKE:
+                        handshake = Handshake.fromByteBuffer(msg);
+                        break;
+                    case TRANSACTION_GOSSIP:
+                        msgsRead++;
+                        txPipeline.process(this, msg);
+                        break;
+                    default:
+                        // do nothing
+                }
+                // reset
+                readState = ReadState.PARSE_HEADER;
+                msgReader = MessageReaderFactory.create(ProtocolMessage.HEADER, ProtocolMessage.HEADER.getMaxLength());
+                return bytesRead;
+            default:
+                // do nothing
+        }
+        return bytesRead;
+    }
+
+    @Override
+    public int write() throws IOException {
+        // previous message wasn't fully sent yet
+        if (currentToWrite != null) {
+            return writeMsg();
+        }
+
+        currentToWrite = sendQueue.poll();
+        if (currentToWrite == null) {
+            return 0;
+        }
+        return writeMsg();
+    }
+
+    private int writeMsg() throws IOException {
+        int written = channel.write(currentToWrite);
+        if (!currentToWrite.hasRemaining()) {
+            msgsWritten++;
+            currentToWrite = null;
+        }
+        return written;
+    }
+
+    @Override
+    public void send(ByteBuffer buf) {
+        // re-register write interest
+        SelectionKey key = channel.keyFor(selector);
+        if (key != null && key.isValid() && (key.interestOps() & SelectionKey.OP_WRITE) == 0) {
+            key.interestOps(SelectionKey.OP_READ | SelectionKey.OP_WRITE);
+            selector.wakeup();
```

Comment:
What do you mean more than once? If you never de-register the `WRITE` interest, than `select()` will continuously execute `write()` on the neighbor because the socket is ready to send stuff. However, if we didn't have anything to send, we can deregister the`WRITE` interest in that moment and only readd the `WRITE` interest when we actually have something to send.